src/workerd/api/node/zlib-util.c++

```diff
@@ -878,6 +836,109 @@ void ZlibUtil::CompressionStream<CompressionContext>::FreeForZlib(void* data, vo
   auto real_pointer = static_cast<uint8_t*>(pointer) - sizeof(size_t);
   JSG_REQUIRE(ctx->allocations.erase(real_pointer), Error, "Zlib allocation should exist"_kj);
 }
+namespace {
+// A RAII wrapper around a compression context class
+template <typename CompressionContext>
+class ContextRAII: public CompressionContext {
+public:
+  using CompressionContext::CompressionContext;
+
+  ~ContextRAII() {
+    static_cast<CompressionContext*>(this)->close();
+  }
+};
+
+template <typename Context>
+static kj::Array<kj::byte> syncProcessBuffer(Context& ctx, GrowableBuffer& result) {
+  do {
+    result.addChunk();
+    ctx.setOutputBuffer(kj::ArrayPtr(result.end(), result.available()));
+
+    ctx.work();
+
+    KJ_IF_SOME(error, ctx.getError()) {
+      JSG_FAIL_REQUIRE(Error, error.message);
+    }
+
+    result.adjustUnused(ctx.getAvailOut());
+  } while (ctx.getAvailOut() == 0);
+
+  return result.releaseAsArray();
+}
+}  // namespace
+
+kj::Array<kj::byte> ZlibUtil::zlibSync(
+    ZlibUtil::InputSource data, ZlibUtil::Options opts, ZlibModeValue mode) {
+  ContextRAII<ZlibContext> ctx;
+
+  auto chunkSize = opts.chunkSize.orDefault(ZLIB_PERFORMANT_CHUNK_SIZE);
+  auto maxOutputLength = opts.maxOutputLength.orDefault(Z_MAX_CHUNK);
+
+  JSG_REQUIRE(Z_MIN_CHUNK <= chunkSize && chunkSize <= Z_MAX_CHUNK, Error, "Invalid chunkSize");
+  JSG_REQUIRE(maxOutputLength <= Z_MAX_CHUNK, Error, "Invalid maxOutputLength");
+  GrowableBuffer result(ZLIB_PERFORMANT_CHUNK_SIZE, maxOutputLength);
+
+  ctx.setMode(static_cast<ZlibMode>(mode));
+  ctx.initialize(opts.level.orDefault(Z_DEFAULT_LEVEL),
+      opts.windowBits.orDefault(Z_DEFAULT_WINDOWBITS), opts.memLevel.orDefault(Z_DEFAULT_MEMLEVEL),
+      opts.strategy.orDefault(Z_DEFAULT_STRATEGY), kj::mv(opts.dictionary));
+  ctx.setFlush(opts.finishFlush.orDefault(Z_FINISH));
+  ctx.setInputBuffer(getInputFromSource(data));
+  return syncProcessBuffer(ctx, result);
+}
+
+void ZlibUtil::zlibWithCallback(
+    jsg::Lock& js, InputSource data, Options options, ZlibModeValue mode, CompressCallback cb) {
+  try {
```

Comment:
These should use `js.tryCatch(...)` instead. Specifically, the `CompressCallback` could end up triggering a `jsg::JsExceptionThrown` which would not be handled by this, and there are fatal exception cases we need to make sure are handled. It would be something like...

Also, there's a risk here that `cb(...)` is invoked twice. It should only ever be invoked exactly once.

Reply:
I think your example got truncated

Reply:
no, but my thought process did lol. I originally intended to write an example but got distracted and forgot to.

Reply:
I think I figured it out - please have a look.